LayoutTests/inspector/console/js-completions.html

```diff
@@ -0,0 +1,54 @@
+<!DOCTYPE html>
+<html>
+<head>
+<script src="../../http/tests/inspector/resources/inspector-test.js"></script>
+<script>
+function test()
+{
+    let suite = InspectorTest.createAsyncSuite("console.jsCompletions");
+
+    suite.addTestCase({
+        name: "console.jsCompletions.variableCompletions",
+        description: "Test that the console prompt's JavaScript completions for variables work as expected.",
+        async test() {
+            const generateJSCompletions = () => new Promise((resolve, reject) => {
+                let mockCompletionController = {
+                    mode: null,
+                    updateCompletions: (completions) => {
+                        resolve(completions);
+                    },
+                };
+
+                const defaultCompletions = [];
+                const baseString = "";
+                const prefix = "";
+                const suffix = "";
+                const forced = true;
+                WI.javaScriptRuntimeCompletionProvider.completionControllerCompletionsNeeded(mockCompletionController, defaultCompletions, baseString, prefix, suffix, forced);
+            });
+
+            let completions = await generateJSCompletions();
+            InspectorTest.expectThat(!completions.includes("myVariable"), "Completions should not contain myVariable by default.");
+
+            await InspectorTest.evaluateInPage(`myVariable = 42;`);
+            WI.javaScriptRuntimeCompletionProvider.clearCachedPropertyNames();
```

Comment:
I'm a slightly concerned that clearing the cache manually will obscure a future bug where the implementation breaks but the test still passes.

Could we instead set `WI.JavaScriptLogViewController.CachedPropertiesDuration = 0` at the beginning of the test and then reset it at the end?

I'm thinking that accelerating time will still test the implementation behavior, but I am concerned that this will introduce flakiness. Pinging @dcrousso for a second opinion.

Reply:
Ah, I too was trying to come up with a way to let the internal code to clear the cache by itself without having to call it here, but I couldn't think of any. Clearing the cache it here manually does look a bit fake.

Is there way for a test to mock running code in inspector's console? (`InspectorTest.evaluateInPage` and `TestPage.evaluateInWebInspector` both don't actually mimic running the code in console and therefore don't trigger `JavaScriptLogViewController.consolePromptTextCommitted`. I thought it'd be best of there's a way to run code that does trigger the said function.)

> Could we instead set `WI.JavaScriptLogViewController.CachedPropertiesDuration = 0` at the beginning of the test and then reset it at the end?
That's an interesting idea! But that's then relying on the completion provider's auto-clearing procedure, still not the console proactively clearing the cache when new code is run, which is the behavior my patch was actually adding 🤔

Reply:
Web Inspector does not have UI tests. If the completions test requires Views to be instantiated or any UI interaction, we can't really test that.

If that's the case, clearing the cache manually to test the controller behavior is fine.

Reply:
OK.

In that case, I think I'd wrap the cache-clearing call into a mock function like this:
```suggestion
            const mockHandleConsolePromptCommitted = () => {
                WI.javaScriptRuntimeCompletionProvider.clearCachedPropertyNames();
            };
            mockHandleConsolePromptCommitted();
```

Just to be clear that this cache-clearing call is supposed to be done by somebody else.

Reply:
It's fine to keep the method inline. Having clear and simple to read tests is good ;)

Reply:
FWIW this test is more about verifying that `WI.javaScriptRuntimeCompletionProvider` works as intended, not necessarily that the hookups to console evaluations are working

as such, id maybe even add some negative tests to verify that things are still (not) in the cache after evaluating until the cache is cleared
```js
let completionsBefore = await generateJSCompletions();
InspectorTest.expectThat(!completions.includes("myVariable"), "Completions should not contain myVariable by default.");

await InspectorTest.evaluateInPage(`myVariable = 42;`);
let completionsAfterEvaluating = await generateJSCompletions();
InspectorTest.expectThat(!completionsAfterEvaluating.includes("myVariable"), "Completions should not contain myVariable after evaluation unless the cache is cleared.");
InspectorTest.assert(Object.shallowEqual(completionsBefore, completionsAfterEvaluating), "Completions should not change unless the cache is cleared.");

WI.javaScriptRuntimeCompletionProvider.clearCachedPropertyNames();
let completionsAfterClearingForEvaluation = await generateJSCompletions();
InspectorTest.expectThat(completionsAfterClearingForEvaluation.includes("myVariable"), "Completions should contain myVariable after it's created by code run in console.");
InspectorTest.assert(!Object.shallowEqual(completionsAfterEvaluating, completionsAfterClearingForEvaluation), "Completions should change after the cache is cleared.");

await InspectorTest.evaluateInPage(`delete myVariable;`);
let completionsAfterDeleting = await generateJSCompletions();
InspectorTest.expectThat(completionsAfterDeleting.includes("myVariable"), "Completions should contain myVariable after deleting unless the cache is cleared.");
InspectorTest.assert(Object.shallowEqual(completionsAfterClearingForEvaluation, completionsAfterDeleting), "Completions should not change unless the cache is cleared.");

WI.javaScriptRuntimeCompletionProvider.clearCachedPropertyNames();
let completionsAfterClearingForDelete = await generateJSCompletions();
InspectorTest.expectThat(!completionsAfterClearingForDelete.includes("myVariable"), "Completions should not contain myVariable after it's deleted by code run in console.");
InspectorTest.assert(!Object.shallowEqual(completionsAfterDeleting, completionsAfterClearingForDelete), "Completions should change after the cache is cleared.");
```